File: 프로그래머스/2/42860. 조이스틱/조이스틱.cpp
#include <string>
#include <vector>

using namespace std;



int GetCount(char ch)
{
    return min(ch  - 65, abs(ch - 91));
}

int solution(string name) {
    int answer = 100000000;

    int result = 0;
    
    vector<int> count(name.length(), 0);
    

    for (int i = 0; i < name.length(); ++i)
    {
        result += GetCount(name[i]);
    }

    if(result == 0)
        return 0;
    
    answer = result + name.length() - 1;

    for (int i = 0; i < name.length(); ++i)
    {
        if (name[i] == 'A')
        {
            int end = i;

            while (end < name.length() && name[end] == 'A')
            {
                ++end;
            }

            answer = min(answer, max((i - 1), 0) * 2 + (int)name.length() - end + result);
            answer = min(answer, ((int)name.length() - end) * 2 + (max(0, i - 1) + result));
            i = end - 1;
        }
    }


    return answer;
}

File: 프로그래머스/2/42860. 조이스틱/조이스틱.java
class Solution {
    public int solution(String name) {
        int answer = 0;
        int result = 0;

        for (int i = 0; i < name.length(); i++) {
            int pos = name.charAt(i) - 'A';
            if (pos <= 13) {
                result += pos;
            } else {
                result += 26 - pos;
            }
        }

        answer = name.length() - 1 + result;
        for (int i = 0; i < name.length(); ++i) {
            if (name.charAt(i) == 'A') {
                int end = i;

                while (end < name.length() && name.charAt(end) == 'A') {
                    ++end;
                }

                answer = Math.min(answer, Math.max(i - 1, 0) * 2 + name.length() - end + result);
                answer = Math.min(answer, (name.length() - end) * 2 + Math.max(i - 1, 0) + result);
                i = end - 1;
            }
        }
        return answer;
    }
}